### PR TITLE
Ted's Implementation

### PR DESCRIPTION
**Pull request by tedmdelacruz**
> What design patterns did you use inside the application?

In the Django backend, I implemented the **Module Pattern**. Django encourages to modularize projects into Django apps and its easy to do so since Python has built in support for modules.  I created two modules (Django apps): `notes_api` and `todo_api`. The URL routes are simply managed in the `api_org_app.urls`. Because of this, the modules are easily pluggable into one another. For example, `notes_api.views` can import the `Todo` model by using `from todos_api.models import Todo`

In the JavaScript, I implemented **Redux** which is inspired by the **Flux Pattern**. Redux manages the app state by listening to synchronous and asynchronous dispatch calls and mutates the state accordingly.

> What libraries did you use?

Django in the backend, and React on Redux in the frontend

> Why did you use this library?

I'm more well versed in the Laravel framework, but I think that framework is too heavy and has too many dependencies. Python is built-in in most systems and can be easily wrapped by virtualenv. Django has a `manage.py` CLI utility which can easily setup the project to run on `localhost:3000`. I want the exam facilitator to easily setup the project without much hassle.

As for React on Redux, I've been studying the technology and I thought it would be fun to use it in the exam.

> Are there other libraries available?

I could have used AngularJS instead of React on Redux

> If yes, what made you choose this library over the others?

I enjoy using React on Redux more than using AngularJS, but I acknowledge that it's not a good enough reason when building real-world applications.

> What major challenges did you encounter when making the app?

I realized that I need to learn more about Redux. As the app became more complex, maintaining the app state became more confusing to me. I had questions like *"Do I have to mutate the state from the within the `<Note/>` component? Or should I create a new dispatch call and handle the state in the Redux `reducer`?".*

> If you had more time, what additional features would you like to add?

1. I would improve some responsive styles.
2. I might have over-engineered the Redux codebase, I would refactor the app further maybe by eliminating the `<App/>` wrapper and use Bootstrap JavaScript to switch between tabs.
3. I would implement React Router and divided the app into two routes: `/notes` and `/todo`
4. I totally forgot to add form validation.

> Other information

The project has been successfully deployed on http://yoyo.tedmdelacruz.com/
Please check the README.md for setup instructions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+## Setup
+    pip install -r requirements.txt
+    npm install
+    bower install
+    gulp
+    python manage.py migrate
+    python manage.py runserver 3000
+
+## Notes
+I might have overengineered the Redux side and violated some Redux patterns.
```
Comment (zakame) [inline README.md]: Yep, very :) I'm mainly interested in the API implementation, not so much in the frontend.

**Comment by zakame** [comment]
Is it primarily a Django feature that you separated the notes and todos into separate modules?  Would a common module for either subclassing or role delegation work here?

**Comment by tedmdelacruz** [comment]
> Is it primarily a Django feature that you separated the notes and todos into separate modules? Would a common module for either subclassing or role delegation work here?

It is a practice encouraged by Django. At first I just had an single API module which manages both Notes and Todos but I realized that the module is doing too many things so I decided to divide it into two modules.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+## Built With
+1. Django Framework 1.9.7
+2. ReactJS on Redux
+3. Sass
+4. Bootstrap
+
+## Setup
+    pip install -r requirements.txt
+    npm install
+    bower install
```
Comment (zakame) [inline README.md]: Ran into a setup issue here:

```
Unable to find a suitable version for jquery, please choose one:
    1) jquery#1.9.1 - 2 which resolved to 2.2.4 and is required by bootstrap#3.3.6
    2) jquery#^3.0.0 which resolved to 3.0.0 and is required by api-org-app

Prefix the choice with ! to persist it to bower.json

? Answer bower error         Invalid regular expression: /.{1,0}/: numbers out of order in {} quantifier.

Stack trace:
SyntaxError: Invalid regular expression: /.{1,0}/: numbers out of order in {} quantifier.
    at new RegExp (native)
    at breakLines (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/inquirer/lib/utils/screen-manager.js:108:15)
    at ScreenManager.render (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/inquirer/lib/utils/screen-manager.js:55:22)
    at Prompt.render (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/inquirer/lib/prompts/input.js:75:15)
    at Prompt._run (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/inquirer/lib/prompts/input.js:49:8)
    at Prompt.run (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/inquirer/lib/prompts/base.js:57:8)
    at Object.<anonymous> (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/inquirer/lib/ui/prompt.js:84:12)
    at module.exports (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/run-async/index.js:15:21)
    at AnonymousObservable.__subscribe (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/inquirer/lib/utils/utils.js:16:7)
    at AnonymousObservable.tryCatcher (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/rx-lite/rx.lite.js:63:31)

Console trace:
Error
    at StandardRenderer.error (/home/zakame/.node_modules/lib64/node_modules/bower/lib/renderers/StandardRenderer.js:83:37)
    at Logger.<anonymous> (/home/zakame/.node_modules/lib64/node_modules/bower/lib/bin/bower.js:110:26)
    at Logger.emit (events.js:107:17)
    at Logger.emit (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/bower-logger/lib/Logger.js:29:39)
    at /home/zakame/.node_modules/lib64/node_modules/bower/lib/commands/index.js:48:20
    at _rejected (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/q/q.js:844:24)
    at /home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/q/q.js:870:30
    at Promise.when (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/q/q.js:1122:31)
    at Promise.promise.promiseDispatch (/home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/q/q.js:788:41)
    at /home/zakame/.node_modules/lib64/node_modules/bower/lib/node_modules/q/q.js:604:44
System info:
Bower version: 1.7.7
Node version: 0.12.7
OS: Linux 4.4.13 x64
```

Manually resolved by forcing resolution to `jquery#^3.0.0` in `bower.json`

Comment (tedmdelacruz) [inline README.md]: I should have indicated `jquery#1.9.1` as required by Bootstrap. If it doesn't work properly with `jquery#3.0.0`, you can simply downgrade jQuery.

**Comment by zakame** [comment]
Seems not working for Python 3?  I'm getting encoding/string issues when creating notes/todos:

```
[21/Jun/2016 07:21:21] "GET /api/todos/ HTTP/1.1" 200 2
Internal Server Error: /api/notes/
Traceback (most recent call last):
  File "/home/zakame/.pyenv/versions/3.5.1/lib/python3.5/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/zakame/.pyenv/versions/3.5.1/lib/python3.5/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/zakame/src/alter-emacs/api-org-app/notes_api/views.py", line 47, in index
    return create_note(request)
  File "/home/zakame/src/alter-emacs/api-org-app/notes_api/views.py", line 16, in create_note
    data = json.loads(request.body)
  File "/home/zakame/.pyenv/versions/3.5.1/lib/python3.5/json/__init__.py", line 312, in loads
    s.__class__.__name__))
TypeError: the JSON object must be str, not 'bytes'
[21/Jun/2016 07:21:34] "POST /api/notes/ HTTP/1.1" 500 83555
Internal Server Error: /api/todos/
Traceback (most recent call last):
  File "/home/zakame/.pyenv/versions/3.5.1/lib/python3.5/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/zakame/.pyenv/versions/3.5.1/lib/python3.5/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/zakame/src/alter-emacs/api-org-app/todos_api/views.py", line 55, in index
    return create_todo(request)
  File "/home/zakame/src/alter-emacs/api-org-app/todos_api/views.py", line 18, in create_todo
    data = json.loads(request.body)
  File "/home/zakame/.pyenv/versions/3.5.1/lib/python3.5/json/__init__.py", line 312, in loads
    s.__class__.__name__))
TypeError: the JSON object must be str, not 'bytes'
[21/Jun/2016 07:22:47] "POST /api/todos/ HTTP/1.1" 500 83526
```

Switching to Python 2.7 makes it work.

**Comment by tedmdelacruz** [comment]
Yes it's not working on Python 3.0 yet

**Comment by zakame** [comment]
LGTM

**Pull request closed without merging**